### PR TITLE
Add a progress bar based on `ProgressMeter.jl`

### LINKED ISSUE #135: Progress report that is a progress bar using `ProgressMeter.jl`
If I understand correctly, currently there are two progress reporters available in DynamicHMC.jl:

| Progress reporter | Outputs | 
| ------------------ | -------- |
| `NoProgressReport` | no output | 
| `LogProgressReport` | prints information every `X` seconds using the `@info` macro |

Would it be possible to add a third progress reporter that is just a simple progress bar using [`ProgressMeter.jl`](https://github.com/timholy/ProgressMeter.jl)?

This would require adding `ProgressMeter.jl` as a dependency. If you are fine with that @tpapp, I can work on making a PR.

I'd like to use this in the [SossMLJ.jl](https://github.com/cscherrer/SossMLJ.jl) package that @cscherrer and I are working on.

### PR DESCRIPTION
**Pull request by DilumAluthge**
Fixes #135

This pull request adds the `ProgressMeterReport`, which uses the `ProgressMeter.jl` package to display a progress bar.

To use this reporter, simply set the `reporter` keyword argument when calling `mcmc_with_warmup`:
```julia
reporter = ProgressMeterReport()
```

cc: @tpapp

### CODE REVIEW

**Comment by DilumAluthge** [comment]
Alright, this is ready for review now @tpapp

**Comment by tpapp** [comment]
Thanks. I am very busy now but will try to do it soon. Geel free to bump this if I don't get to it within a week.

File: src/reporting.jl [hunk 1]
```diff
@@ -34,7 +36,7 @@ Return a reporter which can be used for progress reports with a known number of
 `total_steps`. May return the same reporter, or a related object. Will display `meta` as
 key-value pairs.
 """
-make_mcmc_reporter(reporter::NoProgressReport, total_steps; meta...) = reporter
+make_mcmc_reporter(reporter::NoProgressReport, total_steps; currently_warmup::Bool=false, meta...) = reporter
```
Comment (tpapp) [inline src/reporting.jl:39]:
```suggestion
make_mcmc_reporter(reporter::NoProgressReport, total_steps; currently_warmup::Bool = false, meta...) = reporter
```
(spaces)

File: src/reporting.jl [hunk 2]
```diff
@@ -98,7 +100,7 @@ function report(reporter::LogMCMCReport, message::AbstractString; meta...)
     nothing
 end
 
-function make_mcmc_reporter(reporter::LogProgressReport, total_steps::Integer; meta...)
+function make_mcmc_reporter(reporter::LogProgressReport, total_steps::Integer; currently_warmup::Bool=false, meta...)
```
Comment (tpapp) [inline src/reporting.jl:103]:
```suggestion
function make_mcmc_reporter(reporter::LogProgressReport, total_steps::Integer; currently_warmup::Bool = false, meta...)
```

File: src/reporting.jl [hunk 3]
```diff
@@ -34,7 +36,7 @@ Return a reporter which can be used for progress reports with a known number of
 `total_steps`. May return the same reporter, or a related object. Will display `meta` as
```
Comment (tpapp) [inline src/reporting.jl:36]: please document currently_warmup in the docstring

**Review by tpapp — CHANGES_REQUESTED** [review]
LGTM, I just requested some trivial changes. Thanks for adding this nice feature!

File: src/reporting.jl (hunk 3 above)
Comment (DilumAluthge) [reply src/reporting.jl:36]: Done!

**Comment by DilumAluthge** [comment]
I think I've addressed all the changes!

You will also notice that I have now removed the `log_warmup` field. This bears some explanation.

Right now, we have separate progress bars for warmup and MCMC. Long-term, I think the best solution will be to figure out how to have a single progress bar for both warmup and MCMC. This will require some more info, e.g. https://github.com/tpapp/DynamicHMC.jl/issues/141

Once we implement that feature, the `log_warmup` field will no longer make sense. If we add the `log_warmup` field now, then later, when we are ready to remove it, that would be a breaking change.

So I think it's best to right now remove the `log_warmup` field, and always generate progress bars for both warmup and MCMC. And then in the future we can work on combining the progress bars for warmup and MCMC into a single progress bar.

**Comment by tpapp** [comment]
Thanks, I think this is the best solution for now.

I am merging this, but if that is OK with you I will wait for JuliaNLSolvers/Optim.jl#863 to resolve before tagging a new release. If you need that before, just let me know.

**Comment by DilumAluthge** [comment]
> I am merging this, but if that is OK with you I will wait for [JuliaNLSolvers/Optim.jl#863](https://github.com/JuliaNLSolvers/Optim.jl/issues/863) to resolve before tagging a new release. If you need that before, just let me know.

Sounds good to me, there's no time pressure on my part.

**Pull request merged**